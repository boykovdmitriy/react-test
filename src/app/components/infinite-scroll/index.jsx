import React from 'react';
import PropTypes from 'prop-types';

export class InfiniteScroll extends React.Component {
    constructor(props) {
        super(props);
        this.scrollEventListener = () => this.handleOnScroll();
    }

    componentDidMount() {
        window.addEventListener('scroll', this.scrollEventListener);
    }

    componentWillUnmount() {
        window.removeEventListener('scroll', this.scrollEventListener);
    }

    handleOnScroll() {
        const {handleScroll} = this.props;
        const scrollTop = (document.documentElement && document.documentElement.scrollTop) || document.body.scrollTop;
        const scrollHeight = (document.documentElement && document.documentElement.scrollHeight) || document.body.scrollHeight;
        const clientHeight = document.documentElement.clientHeight || window.innerHeight;
        const scrolledToBottom = Math.ceil(scrollTop + clientHeight) >= scrollHeight;

        if (scrolledToBottom) {
            handleScroll();
        }
    };

    render() {
        const {children} = this.props;
        return (
            <section>
                {children}
            </section>
        )
    }
}

InfiniteScroll.propTypes = {
    children: PropTypes.object.isRequired,
    handleScroll: PropTypes.func.isRequired
};