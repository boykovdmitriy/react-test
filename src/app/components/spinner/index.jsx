import React from 'react'
import style from './spinner.css';

export const Spinner = () => (
    <div className={style.loader}/>
);