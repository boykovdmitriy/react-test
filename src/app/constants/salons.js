import keyMirror from 'fbjs/lib/keyMirror';

export const Salons = keyMirror({
    SALON_REQUEST: undefined,
    SALON_SUCCESS: undefined,
    SALON_FAILURE: undefined,
    SALON_CANCEL: undefined,
    SALONS_REQUEST: undefined,
    SALONS_SUCCESS: undefined,
    SALONS_FAILURE: undefined,
    SALONS_CANCEL: undefined,
});