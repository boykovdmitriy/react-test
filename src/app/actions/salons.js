import {Salons} from "constants";

export function salonsRequest(page, pageSize) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {page, pageSize},
    };
}

export function salonRequest(salonId) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {salonId},
    };
}