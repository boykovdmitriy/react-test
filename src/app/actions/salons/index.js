import {Salons} from "constants/salons";

export function salonsRequest(page, pageSize) {
    return {
        type: Salons.SALONS_REQUEST,
        payload: {page, pageSize},
    };
}

export function salonRequest(salonId) {
    return {
        type: Salons.SALON_REQUEST,
        payload: {salonId},
    };
}