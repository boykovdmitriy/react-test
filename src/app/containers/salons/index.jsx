import React from "react";
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import {InfiniteScroll} from 'components/infinite-scroll';

import {ProfileImage} from "components/profile-image";
import {salonsRequest} from "actions";
import {Table, TableColumn, TableRow} from "components/table";
import {Spinner} from 'components/spinner';

class Salons extends React.Component {
    constructor(props) {
        super(props);
        this.pageSize = 25;
    }

    componentDidMount() {
        const {loadedUntilPage} = this.props;
        if (!loadedUntilPage) {
            this.loadNextPage();
        }
    }

    loadNextPage() {
        const {fetchSalons, loadedUntilPage, total} = this.props;
        const nextPage = loadedUntilPage ? loadedUntilPage + 1 : 1;
        if (!total || nextPage <= total) {
            fetchSalons(nextPage, this.pageSize);
        }
    }

    handleScroll() {
        this.loadNextPage();
    }

    render() {
        const {isLoading, salons} = this.props;
        return (
            <section>
                <InfiniteScroll handleScroll={() => this.handleScroll()}>
                    <Table headers={['Name', 'Website', 'Image']}>
                        {salons.length !== 0 && salons.map(x => (
                            <TableRow key={x.id}>
                                <TableColumn>{x.name}</TableColumn>
                                <TableColumn>{!!x.website && <a href={x.website}>link</a>}</TableColumn>
                                <TableColumn>{
                                    !!x.profile_image_urls &&
                                    <ProfileImage thumb={x.profile_image_urls.thumb}/>
                                }</TableColumn>
                                <TableColumn><Link to={`salon/${x.id}`}>Details</Link></TableColumn>
                            </TableRow>
                        ))}
                    </Table>
                </InfiniteScroll>
                {isLoading && <Spinner/>}
            </section>
        );
    }
}

const mapStateToProps = state => ({
    isLoading: state.salons.list.isLoading,
    salons: state.salons.list.items,
    totalPage: state.salons.list.total,
    loadedUntilPage: state.salons.list.page
});
const mapDispatchToProps = dispatch => ({
    fetchSalons: (page, pageSize) => dispatch(salonsRequest(page, pageSize))
});
export default connect(mapStateToProps, mapDispatchToProps)(Salons);



