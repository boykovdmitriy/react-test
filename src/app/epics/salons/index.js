import {ofType} from 'redux-observable';
import {switchMap} from 'rxjs/operators'

import {Salons} from 'constants/salons';
import {rxAjax} from 'utils';

export function salonsRequest(action$) {
    return action$.pipe(
        ofType(Salons.SALONS_REQUEST),
        switchMap((action) => {
            const endpoint = `http://staging.salony.com/v5/salons?per_page=${action.payload.pageSize}&page=${action.payload.page}`;

            /* istanbul ignore next */
            return rxAjax({
                endpoint,
                method: 'GET',
            })
                .map(data => ({
                    type: Salons.SALONS_SUCCESS,
                    payload: data.response,
                }))
                .catch(error => ({
                    type: Salons.SALONS_FAILURE,
                    payload: {message: error.error, status: error.status},
                }));
        })
    );
}

export function salonRequest(action$) {
    return action$.pipe(
        ofType(Salons.SALON_REQUEST),
        switchMap((action) => {
            const endpoint = `http://staging.salony.com//v5/salons/${action.payload.salonId}`;

            /* istanbul ignore next */
            return rxAjax({
                endpoint,
                method: 'GET',
            })
                .map(data => ({
                    type: Salons.SALON_SUCCESS,
                    payload: data.response,
                }))
                .catch(error => ({
                    type: Salons.SALON_FAILURE,
                    payload: {message: error.error, status: error.status},
                }));
        })
    );
}
