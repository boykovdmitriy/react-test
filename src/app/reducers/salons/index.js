import {Salons} from "constants/salons";

export const initialState = {
    list: {
        isLoading: false,
        items: [],
    },
    item: {
        isLoading: false,
    }
};

export const reducer = (state = initialState, action) => {
    switch (action.type) {
        case Salons.SALONS_REQUEST:
            return {
                ...state,
                list: {
                    ...state.list,
                    isLoading: true,
                    page: action.payload.page
                },
            };
        case Salons.SALONS_SUCCESS:
            return {
                ...state,
                list: {
                    ...state.list,
                    items: [...state.list.items, ...action.payload.salons],
                    total: action.payload.total,
                    isLoading: false,
                },
            };
        case Salons.SALONS_FAILURE:
            return {
                ...state,
                list: {
                    ...state.list,
                    error: action.payload,
                    isLoading: false,
                }
            };
        case Salons.SALON_REQUEST:
            return {
                ...state,
                item: {
                    isLoading: true,
                }
            };
        case Salons.SALON_SUCCESS:
            return {
                ...state,
                item: {
                    isLoading: false,
                    ...action.payload
                }
            };
        case Salons.SALON_FAILURE:
            return {
                ...state,
                item: {
                    isLoading: false,
                    error: action.payload
                }
            };
        default:
            return state;
    }
};