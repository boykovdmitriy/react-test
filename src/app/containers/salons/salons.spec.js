import React from 'react';
import configureMockStore from 'redux-mock-store';
import {Provider} from 'react-redux';
import {MemoryRouter} from 'react-router-dom';

import ConnectedSalons, {Salons} from './index'
import {Salons as SalonsConstants} from "constants/salons";
import {Spinner} from 'components/spinner';
import {InfiniteScroll} from 'components/infinite-scroll';


const initialState = {
    salons: {
        list: {
            isLoading: false,
            items: [{
                id: 1,
                name: 'first',
                profile_image_urls: {thumb: 'http://s3-eu-west-1.amazonaws.com/staging-salony/image/salons/files/000/003/623/medium/nail-art-2392-250x250.jpg?1507118077'}
            }, {
                id: 2,
                name: 'second',
                profile_image_urls: {thumb: 'http://s3-eu-west-1.amazonaws.com/staging-salony/image/salons/files/000/003/623/medium/nail-art-2392-250x250.jpg?1507118077'}
            }],
            page: 2,
            total: 12
        }
    },
};
describe('Salons shallow and passing store directly', () => {
    let store, container;
    const mockStore = configureMockStore();

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(<ConnectedSalons store={store}/>)
    });

    it('render Salons component', () => {
        const wrapper = shallow(
            <Salons/>
        );
        expect(wrapper).toMatchSnapshot();
    });

    it('render connected(Salons) component', () => {
        expect(container.length).toEqual(1)
    });
});

describe('Salons mount and wrapping in Provider', () => {
    let store, wrapper;
    const mockStore = configureMockStore();

    beforeEach(() => {
        store = mockStore(initialState);
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <ConnectedSalons/>
                </MemoryRouter>
            </Provider>)
    });

    it('render connected(SalonDetails) component', () => {
        expect(wrapper.length).toEqual(1)
    });

    it('check Props matches with initialState', () => {
        store.subscribe(() => {
            expect(wrapper.find(ConnectedSalons).prop('isLoading')).toEqual(initialState.salons.list.isLoading);
            expect(wrapper.find(ConnectedSalons).prop('salons')).toEqual(initialState.salons.list.items);
            expect(wrapper.find(ConnectedSalons).prop('totalPage')).toEqual(initialState.salons.list.total);
            expect(wrapper.find(ConnectedSalons).prop('loadedUntilPage')).toEqual(initialState.salons.list.page);
        });
    });

    it('check handleScroll', () => {
        const pageSize = 25;
        const page = initialState.salons.list.page;
        const handleScroll = wrapper.find(InfiniteScroll).prop('handleScroll');
        store.subscribe(() => {
            handleScroll();
            expect(store.getActions()).toEqual([
                {type: SalonsConstants.SALONS_REQUEST, payload: {page, pageSize}},
            ]);
        });
    });

    it('check actions on dispatching', () => {
        const pageSize = 25;
        const page = initialState.salons.list.page;
        store.subscribe(() => {
            const instance = wrapper.find(ConnectedSalons).children().instance();
            const fetchSalons = wrapper.find(ConnectedSalons).prop('fetchSalons');
            fetchSalons();
            instance.loadNextPage();
            expect(store.getActions()).toEqual([
                {type: SalonsConstants.SALONS_REQUEST, payload: {page, pageSize}},
                {type: SalonsConstants.SALONS_REQUEST, payload: {page: page + 1, pageSize}},
            ]);
        });
    });
});

describe('Salons mount and wrapping in Provider for testing spinner', () => {
    let store, wrapper;
    const mockStore = configureMockStore();

    beforeEach(() => {
        store = mockStore({
            salons: {
                list: {
                    isLoading: true
                }
            }
        });
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <ConnectedSalons/>
                </MemoryRouter>
            </Provider>)
    });

    it('render connected(SalonDetails) component', () => {
        expect(wrapper.length).toEqual(1)
    });

    it('check Spinner while loading', () => {
        expect(wrapper.contains(<Spinner/>)).toEqual(true);
    });
});