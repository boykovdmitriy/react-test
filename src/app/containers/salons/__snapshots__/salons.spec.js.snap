// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Salons shallow and passing store directly render Salons component 1`] = `
<section>
  <InfiniteScroll
    handleScroll={[Function]}
  >
    <Table
      headers={
        Array [
          "Name",
          "Website",
          "Image",
        ]
      }
    />
  </InfiniteScroll>
</section>
`;
